api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java

```diff
@@ -0,0 +1,40 @@
+package net.md_5.bungee.api.event;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import net.md_5.bungee.api.ChatMessageType;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.connection.Connection;
+import net.md_5.bungee.api.plugin.Cancellable;
+import net.md_5.bungee.chat.ComponentSerializer;
+
+/**
+ * Event called when a server sends a message to a player.
+ */
+@Data
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class ServerChatEvent extends TargetedEvent implements Cancellable
+{
+
+    /**
+     * Cancelled state.
+     */
+    private boolean cancelled;
+    /**
+     * Text contained in this chat.
+     */
+    private BaseComponent[] message;
+    /**
+     * Type of message/display location
+     */
+    private ChatMessageType position;
+
+    public ServerChatEvent(Connection sender, Connection receiver, String json, byte position)
+    {
+        super( sender, receiver );
+        this.message = ComponentSerializer.parse( json );
```

Comment:
This value _could_ be computed lazy, not sure if its worth to go for it.

Reply:
Ideally if nothing ever edits the value, we should never need to parse and re-serialise it, but I don't think there's a way to check that aside from going through the getter

Reply:
Just because a plugin fetched the message doesn't mean that it edited it. Most of the time a plugin will have to call the getter just to view the message even without the intent to edit it. The only way I can think is to override every setter in the chat.api.* class to check for modification

Reply:
Just parse the message if someone uses it.

```diff
diff --git a/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java b/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
index b9f54989..e21464d3 100644
--- a/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
@@ -1,7 +1,11 @@
 package net.md_5.bungee.api.event;
 
+import java.util.function.Supplier;
+
 import lombok.Data;
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
 import lombok.ToString;
 import net.md_5.bungee.api.ChatMessageType;
 import net.md_5.bungee.api.chat.BaseComponent;
@@ -12,7 +16,6 @@ import net.md_5.bungee.chat.ComponentSerializer;
 /**
  * Event called when a server sends a message to a player.
  */
-@Data
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class ServerChatEvent extends TargetedEvent implements Cancellable
@@ -21,20 +24,35 @@ public class ServerChatEvent extends TargetedEvent implements Cancellable
     /**
      * Cancelled state.
      */
+    @Getter
+    @Setter
     private boolean cancelled;
     /**
      * Text contained in this chat.
      */
-    private BaseComponent[] message;
+    @Setter
+    private BaseComponent[] message = null;
+    private Supplier<BaseComponent[]> messageSupplier;
     /**
      * Type of message/display location
      */
+    @Getter
+    @Setter
     private ChatMessageType position;
 
     public ServerChatEvent(Connection sender, Connection receiver, String json, byte position)
     {
         super( sender, receiver );
-        this.message = ComponentSerializer.parse( json );
+        this.messageSupplier = () -> ComponentSerializer.parse( json );
         this.position = ChatMessageType.values()[position];
     }
+
+    public BaseComponent[] getMessage() {
+        if (messageSupplier != null) {
+            message = messageSupplier.get();
+            messageSupplier = null;
+        }
+
+        return message;
+    }
 }

```

Reply:
That's fine if someone replaces the entire message object, but it wouldn't detect a change internally to one of the `*Component` objects- or should we treat them as if they are immutable?

Reply:
> That's fine if someone replaces the entire message object, but it wouldn't detect a change internally to one of the *Component objects- or should we treat them as if they are immutable?

Ah you are right Components are mutable; Then this approach is not feasible.

Reply:
Agreed. Applied